File: lintcode/18_subsets-ii/subsets-ii.cpp
/*
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/subsets-ii
@Language: C++
@Datetime: 16-04-24 18:52
*/

class Solution {
public:
    /**
     * @param S: A set of numbers.
     * @return: A list of lists. All valid subsets.
     */
    vector<vector<int> > subsetsWithDup(const vector<int> &S) {
        vector<vector<int> > res;
        vector<int> one;
        vector<int> nums(S);
        sort(nums.begin(), nums.end());
        helper(nums, 0, nums.size(), one, res);
        return res;
    }
    void helper(const vector<int> &nums, int p, int n, vector<int> &one, vector<vector<int> > &res) {
        res.push_back(one);
        for (int i = p; i < n; ++i) {
            if (i > p && nums[i] == nums[i-1]) continue;
            one.push_back(nums[i]);
            helper(nums, i+1, n, one, res);
            one.pop_back();
        }
    }
};

File: lintcode/18_subsets-ii/subsets-ii.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   chenweikang
@Problem:  http://www.lintcode.com/problem/subsets-ii
@Language: Python
@Datetime: 16-07-06 00:09
'''

class Solution:
    """
    @param S: A set of numbers.
    @return: A list of lists. All valid subsets.
    """
    def subsetsWithDup(self, S):
        n = len(S)
        one = []
        res = []
        S.sort()
        def dfs(p):
            res.append(one[:])
            for i in range(p, n):
                if i > p and S[i] == S[i-1]:
                    continue
                one.append(S[i])
                dfs(i+1)
                one.pop()
        dfs(0)
        return res